### PR TITLE
Load dynamic module (remote code) only once if code isn't change

### PR DESCRIPTION
**Pull request by XuehaiPan**
# What does this PR do?

Fixes https://github.com/huggingface/transformers/pull/30370#discussion_r1733458884

Add an indicator `__transformers_module_hash__` to the remote code module.

## Before submitting
- [X] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ydshieh

### CODE REVIEW

**Comment by ydshieh** [comment]
Hi @XuehaiPan, thank you for taking the action.

My first question is: what if the target (remote code) is changed since the last time it has been loaded.

I don't know very well what is the context mentioned in https://github.com/huggingface/transformers/pull/30370#discussion_r1733458884, but IIRD, that happens within a single python process (i.e. multiple loading of the same module). And if that is the case, my question on top of this comment have to be considered, right?

**Comment by XuehaiPan** [comment]
> My first question is: what if the target (remote code) is changed since the last time it has been loaded.

@ydshieh Thanks for raising this. I change this indicator to the hash of the source code.

**Comment by ydshieh** [comment]
Nice. I still prefer for @tmm1 to elaborate the issue a bit more though.


But in the meantime, could you tell me more about the usage of `threading.Lock()` in this PR 🙏 ?

**Comment by XuehaiPan** [comment]
> But in the meantime, could you tell me more about the usage of `threading.Lock()` in this PR 🙏 ?

It's just for thread safety. We are modifying the global variable `sys.modules`.

**Comment by tmm1** [comment]
> prefer for @tmm1 to elaborate the issue a bit more though

sure, here is a simple example:

```python
import sys
from accelerate import init_empty_weights
from transformers import AutoModelForCausalLM

def load_modeling_code(model_name):
    with init_empty_weights():
        model = AutoModelForCausalLM.from_pretrained(
            model_name, trust_remote_code=True
        )
        return sys.modules[model.__class__.__module__]

model_name = "deepseek-ai/DeepSeek-Coder-V2-Lite-Base"

# 1. patch modeling code
mod = load_modeling_code(model_name)
import liger_kernel
mod.CrossEntropyLoss = liger_kernel.transformers.cross_entropy.LigerCrossEntropyLoss
mod.DeepseekV2RMSNorm = liger_kernel.transformers.rms_norm.LigerRMSNorm
mod.DeepseekV2MLP.forward = liger_kernel.transformers.swiglu.LigerSwiGLUMLP.forward

# 2. check patch
mod = load_modeling_code(model_name)
print(mod.CrossEntropyLoss == liger_kernel.transformers.cross_entropy.LigerCrossEntropyLoss)

# 3. create and use model
model = AutoModelForCausalLM.from_pretrained(model_name, trust_remote_code=True)
```

currently on main branch, the patch is missing in step (2) and (3)

**Comment by XuehaiPan** [comment]
It looks reasonable to assume the import statement returns the cached module object if it has already been imported.

Ideally, we may need to simulate this:

```python
sys.path.insert(HF_MODULES_CACHE)
module = importlib.import_module(name)
assert sys.path[0] == HF_MODULES_CACHE
del sys.path[0]
```

**Comment by ydshieh** [comment]
Hi @tmm1 Thank you for providing the detailed information.

I am wondering if it would make sense for the patching code to have a factory method like

```python
def load_modeling_code(model_name):
    with init_empty_weights():
        model = AutoModelForCausalLM.from_pretrained(
            model_name, trust_remote_code=True
        )
        return sys.modules[model.__class__.__module__]

def patch_modeling_code(model_name):

    # 1. patch modeling code
    mod = load_modeling_code(model_name)
    import liger_kernel
    mod.CrossEntropyLoss = liger_kernel.transformers.cross_entropy.LigerCrossEntropyLoss
    mod.DeepseekV2RMSNorm = liger_kernel.transformers.rms_norm.LigerRMSNorm
    mod.DeepseekV2MLP.forward = liger_kernel.transformers.swiglu.LigerSwiGLUMLP.forward

    return mod

model_name = "deepseek-ai/DeepSeek-Coder-V2-Lite-Base"
mod = patch_modeling_code(model_name)
```

But I am open to what @XuehaiPan have done so far.

**Comment by LysandreJik** [comment]
cc @Rocketknight1 as well :)

**Comment by tmm1** [comment]
> I am wondering if it would make (more) sense for the patching code to have a factory method like

it could work, but often the user wants to apply some patch then invoke a well known trainer framework. such a framework would not accept a factory callback

more importantly, the factory would still need to call `AutoModelForCausalLM.from_pretrained` two times:
1. to download and load the remote code, so it is available inside sys.modules for patching
2. without `init_empty_weights()`, and to [actually use the patched code](https://github.com/linkedin/Liger-Kernel/pull/170)

without this PR, the second invocation will remove the patch. so it would be impossible to get a `model` object that actually is patched.

**Comment by ydshieh** [comment]
OK, I didn't know
```
# 3. create and use model
model = AutoModelForCausalLM.from_pretrained(model_name, trust_remote_code=True)
```
in your previous comment is actually needed. Makes sense now.

I will let @Rocketknight1 to have a look on the changes of this PR too.

**Comment by Rocketknight1** [comment]
It seems clean to me! I can think of two failure cases:

1) This will not detect changes in other files (for example, if the model also has utils/functions in a separate file from the main modeling file, the hash of the main modeling file will not change and so it will not be reimported.
2) I think this could cause a regression if users load a remote_code model, monkey-patch methods, and then reload the model to clear their changes, since now their changes will persist. However, I suspect there are no users depending on this weird behaviour, so it's not a serious problem.

Both of these are very minor issues that cannot really be fixed in this PR, and I don't think they should block it, so I'm happy with it!

**Comment by ydshieh** [comment]
Hi @XuehaiPan It would be great if you can add a corresponding test, like a (simple version of) patched module (like what @tmm1 has) will remain the same.

**Comment by XuehaiPan** [comment]
> It would be great if you can add a corresponding test

Done.

File: src/transformers/dynamic_module_utils.py
Comment (ydshieh) [inline src/transformers/dynamic_module_utils.py:232]: we don't need `force_reload` here I believe, as we already use `sys.modules.pop(name, None)` above

File: tests/models/auto/test_feature_extraction_auto.py
Comment (ydshieh) [inline tests/models/auto/test_feature_extraction_auto.py:131]: here is not about `configuration`

File: tests/models/auto/test_tokenization_auto.py
Comment (ydshieh) [inline tests/models/auto/test_tokenization_auto.py:323]: no need to check `self.assertTrue(reloaded_tokenizer.special_attribute_present)`

File: tests/models/auto/test_tokenization_auto.py
Comment (ydshieh) [inline tests/models/auto/test_tokenization_auto.py:361]: yes, here we need to check `self.assertTrue(reloaded_tokenizer.special_attribute_present)`. Nice

**Review by ydshieh — COMMENTED** [review]
very nice. a few nit comments

**Comment by ydshieh** [comment]
In my previous comment

```
        # The configuration file is cached in the snapshot directory. So the module file is not changed after dumping
```
in other files (other than `tests/models/auto/test_configuration_auto.py`)

> here is not about configuration

What I mean is that we have to change the **comment** `The configuration file` to the actual file name (like the image processor file). Otherwise, that part of test is actually ok.

File: src/transformers/dynamic_module_utils.py
Comment (ydshieh) [inline src/transformers/dynamic_module_utils.py:240]: typo? maybe hash hit ..?

**Comment by XuehaiPan** [comment]
> I can go with the current version, just let me know if you want to add it back (with the correct comments).

@ydshieh I add a commit to address this.

**Review by ydshieh — APPROVED** [review]

**Comment by ydshieh** [comment]
@Rocketknight1  in case if you want to take a final look. (and/or feel free to ping a core maintainer 🙏 )

**Comment by Rocketknight1** [comment]
I'm happy with it, I think! The test failure seems unrelated, and I like the core goal of allowing the same model to be imported twice without wasting time, and without getting two different output classes.

cc @XuehaiPan maybe rebase onto `main` and see if that fixes the test?

cc @LysandreJik for core maintainer approval!

**Review by LysandreJik — APPROVED** [review]
Thanks! This looks good to me if approved by the esteemed @Rocketknight1

**Comment by Rocketknight1** [comment]
@XuehaiPan let me know if you're happy to merge, or if there's anything else you want to tweak first!

**Comment by XuehaiPan** [comment]
> @XuehaiPan let me know if you're happy to merge, or if there's anything else you want to tweak first!

@Rocketknight1 I think this is the final version of the PR.

**Comment by Rocketknight1** [comment]
Merging, in that case. Thank you for the PR!

**Pull request merged**
